libsolidity/analysis/ContractLevelChecker.cpp

```diff
@@ -458,3 +461,16 @@ void ContractLevelChecker::checkPayableFallbackWithoutReceive(ContractDefinition
 				SecondarySourceLocation{}.append("The payable fallback function is defined here.", fallback->location())
 			);
 }
+
+void ContractLevelChecker::checkStateVariablesOffsets(ContractDefinition const& _contract)
+{
+	try
+	{
+		// dry run, to make sure it succeeds later
```

Comment:
To me this still feels pretty hacky and it still isn't particularily nice that it throws and catches exceptions here...
There is really no need to have this check here be absolutely exact - it would be enough to use some estimate like ``StaticAnalyzer::structureSizeEstimate`` (which we would move from ``StaticAnalyzer`` to ``Type`` and could later merge with ``validForLocation``).

If that size estimate is small enough (i.e. ``< 2^256/32``) we can be sure that the actual packed storage offsets computation will succeed as well and just assert there.

The estimate not being exact is no issue, since a contract with *roughly* ``2^251`` storage slots is equally undeployable as one with *exactly* ``2^251`` storage slots.

Reply:
Having the check directly in the static analyzer like I tried here? https://github.com/ethereum/solidity/pull/9129

Reply:
@ekpyron 
I completely agree that throwing/catching as a part of a regular workflow is not nice. But I thought that it might be a smaller evil. Although checks are not required to be exact, having them significantly different does not feel quite right too. We sort of modeling during stage A what we'd need later during stage B.

Anyway, regarding your suggestion, should it be implemented around `StaticAnalyzer::structureSizeEstimate()` or `Type::storageSize()`?

Reply:
I think it makes sense to move ``StaticAnalyzer::structureSizeEstimate`` to ``Type::structureSizeEstimate``. It is pretty similar to what ``ReferenceType::validForLocation`` does, so I'd hope that we could merge the logic in a second step.

The logic wouldn't be "entirely different" - you would just not think of state variables to be packed, i.e. ``contract C { uint8 a; uint8 b; }`` would be overcounted as two slots, even though we'd pack it into one.
But that's fine. This is only really relevant for storage sizes near 2^256 - and in these magnitudes an estimation error of a factor of 32 max won't make a difference :-). Nobody will complain if a contract of size ``2^251`` for which it might be possible to pack it into less than ``2^256`` bytes of storage won't compile, because nothing like that can ever be deployed anyways :-).

Reply:
The only thing we need to make sure is that ``structureSizeEstimate`` *over*-estimates and not *under*-estimates, but that it should already do, resp. should be straightforward to achieve.